### PR TITLE
feat: implement v2 discovery API

### PR DESCRIPTION
**Pull request by hahattan**
Signed-off-by: Chris Hung <chris@iotechsys.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://wiki.edgexfoundry.org/display/FA/Committing+Code+Guidelines#CommittingCodeGuidelines-CommitMessages
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

### CODE REVIEW

**Review by lenny-intel — DISMISSED** [review]
LGTM, but swagger indicates a `500` status, which I don't see a case for. Should this `500` status be removed from the swagger?

**Comment by cloudxxx8** [comment]
@lenny-intel I would prefer to keep `500` status code.  It covers all unknown internal error, and we might reply it in the future.
On the other hand, the Swagger API definition is for all kinds of Device Service implementation, so `500` might be returned by device services implemented in other technologies.

**Comment by hahattan** [comment]
rebased

**Comment by lenny-intel** [comment]
> I would prefer to keep `500` status code. It covers all unknown internal error, and we might reply it in the future.
> On the other hand, the Swagger API definition is for all kinds of Device Service implementation, so `500` might be returned by device services implemented in other technologies.

@hahattan , just note that TAF team will use the swagger to determine error tests and will expect to create a Test Case that triggers the 500 status code. Also, the SDks (Go & C) should be consistent on what triggers each status code, i.e. C SDK shouldn't return a 500 when the Go SDK returns something else for the same input.

**Review by lenny-intel — APPROVED** [review]

**Comment by cloudxxx8** [comment]
> > I would prefer to keep `500` status code. It covers all unknown internal error, and we might reply it in the future.
> > On the other hand, the Swagger API definition is for all kinds of Device Service implementation, so `500` might be returned by device services implemented in other technologies.
> 
> @cloudxxx8 , just note that TAF team will use the swagger to determine error tests and will expect to create a Test Case that triggers the 500 status code. Also, the SDks (Go & C) should be consistent on what triggers each status code, i.e. C SDK shouldn't return a 500 when the Go SDK returns something else for the same input.

@lenny-intel the QA team could only test 4XX negative test cases.  5XX should not be produced by the requests.

File: internal/v2/controller/http/discovery.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	ds := container.DeviceServiceFrom(c.dic.Get)
+	correlationID := request.Header.Get(sdkCommon.CorrelationHeader)
+
+	err := checkServiceLocked(request, ds.AdminState)
+	if err != nil {
+		c.sendError(writer, request, edgexErr.KindServiceLocked, "Service Locked", err, sdkCommon.APIV2DiscoveryRoute, correlationID)
+		return
+	}
+
+	if ds.OperatingState == contract.Disabled {
```
Comment (cloudxxx8) [inline internal/v2/controller/http/discovery.go:23]: In V2 API, we decided to remove OperatingState from Device Service, so please remove this section

**Review by cloudxxx8 — CHANGES_REQUESTED** [review]

**Review by cloudxxx8 — APPROVED** [review]
LGTM

**Comment by cloudxxx8** [comment]
> @cloudxxx8 , just note that TAF team will use the swagger to determine error tests and will expect to create a Test Case that triggers the 500 status code. Also, the SDks (Go & C) should be consistent on what triggers each status code, i.e. C SDK shouldn't return a 500 when the Go SDK returns something else for the same input.

If the error is caused by any middleware layer, such as header manipulating, or third party http library, such as gorilla mux although possibility is small, it would become 500 and we can't control it.  I thought we should always leave 500 internal error in each API definition.

**Pull request merged**